### PR TITLE
basic support for importing .xcframeworks

### PR DESCRIPTION
**Pull request by xiemotongye**
This pull request is related to https://github.com/bazelbuild/rules_apple/pull/1166.

The original pull request did a great job, but it seems that there is no follow-up for a long time.

Now I'm facing the same thing, I hope rules_apple can officially support for the xcframwork import. So I did a little work on it.

Usage example:

```
apple_dynamic_framework_import(
    name = "AmazonChimeSDK",
    framework_imports = glob(["AmazonChimeSDK.xcframework/**"]),
    visibility = ["//visibility:public"],
    xcframework_platform_ids = {
        "iphoneos": "ios-arm64_armv7",
        "iphonesimulator": "ios-i386_x86_64-simulator",
    },
)
```

### CODE REVIEW

**Review by thii — APPROVED** [review]
It would be nice if you can add a test.

File: apple/internal/apple_framework_import.bzl
Comment (brentleyjones) [inline apple/internal/apple_framework_import.bzl:324]: @thii Because of the known bug with `ctx.fragments.apple.single_arch_platform`, we should patch this similar to https://github.com/bazelbuild/rules_swift/pull/694, after this is merged. Then we can revert both patches when Bazel core is fixed.

File: apple/internal/apple_framework_import.bzl
Comment (brentleyjones) [inline apple/internal/apple_framework_import.bzl:501]: I guess I haven't thought through this enough, but is this something that needs to be specified, versus Bazel figuring out itself?

Comment (thii) [reply apple/internal/apple_framework_import.bzl:501]: Because Bazel cannot read files at this phase https://github.com/bazelbuild/rules_apple/issues/851#issuecomment-663770260

We could move everything to the tool so that the detection would happen in the execution phase, but that would add more complexity and the imported files would be opaque to Bazel.

All other rules here also just defer the `Info.plist` reading to the rules consumers https://github.com/bazelbuild/rules_apple/blob/54e906af5a45ecd6863bc73d60c7a7419d01bea7/doc/common_info.md#variable-substitution

File: apple/internal/apple_framework_import.bzl
Comment (thii) [reply apple/internal/apple_framework_import.bzl:324]: The bug is with `single_arch_cpu`, not `single_arch_platform`.

**Comment by keith** [comment]
note this approach doesn't solve consuming static xcframeworks that are themselves produced by rules_apple

File: apple/internal/apple_framework_import.bzl
Comment (zachgrayio) [reply apple/internal/apple_framework_import.bzl:501]: Yeah FWIW i did prototype doing all this automatically, and hit some big blockers like @thii mentioned. The internals need the `File` handles and if you do it more dynamically you're left with just the opaque `TreeArtifact` and it breaks most of the functionality. Still have that commit around if anyone wants to see it again--not super clean but you can see what I was working towards. https://github.com/zachgray-slack/rules_apple/commit/201c0577eaab79e65347164bcb8d1f571d8d3115

**Comment by zachgrayio** [comment]
@keith now rules_apple is producing xcframeworks im happy to TAL at this again and find a solution that works for consuming these as well. I'm actually looking at this stuff again as it turns out, maybe we can get something landed this time lol

**Review by zachgrayio — CHANGES_REQUESTED** [review]
so at a high level, this LGTM as it's an improvement on my work, but i guess there's 2 blockers:

1) needs a test probably, something I didn't get to either lol
2) if this patch doesnt support the xcframeworks this ruleset itself creates, we should definitely address that. When I created my draft, this stuff wasn't landed at all yet, so I didn't factor it in. if @keith can share more info here maybe we can figure out what changes we'd need to support that (or i can run a few quick experiments).

FWIW this same patch is applied manually in a few orgs and definitely in use in production, so it does work pretty well for the majority of cases...however we can assume it will quickly become an issue with upcoming Google releases of .xcframeworks that this won't support 😆

File: apple/internal/apple_framework_import.bzl
Comment (erikkerber) [inline apple/internal/apple_framework_import.bzl:341]: Might want `path` here over `short_path`, which will return a relative path when the framework imports are in a different repo.

**Comment by keith** [comment]
The issue I saw is that when you produce a static binary xcframework it doesn't use frameworks, it uses static libs directly, so this doesn't work in general for that. It also makes me wonder if these rules should be used at all for importing xcframeworks, or if there should be separate rules for that. It sounds like static framework xcframework support is spotty in other places, which is why it wasn't implemented in the bazel rule, I haven't tested it but I guess swiftpm doesn't support that case.

**Comment by thii** [comment]
We could add support for that too. I think xcframeworks with static libs aren't that common so no one has asked for it.

**Comment by keith** [comment]
Yea you just don't have another option if you want to support the new sim and statically link things in general. But that made me wonder if these rules weren't the right ones to bake this in to

**Comment by thii** [comment]
@xiemotongye Do you mind if I finish this?

**Comment by thii** [comment]
@xiemotongye Pinging again - if you see this, can you update your PR to allow maintainers to push to your branch? Thanks.

**Comment by thii** [comment]
Submitted https://github.com/bazelbuild/rules_apple/pull/1344 that also supports for static libs, and doesn't need to pass platform ids.

**Comment by xiemotongye** [comment]
> @xiemotongye Do you mind if I finish this?

Sorry for reply so late, I was just returning from vacation of Chinese New Year. And I'm very pleasure that you've finished this, buddy!

**Pull request closed without merging**
